fs/operations/resume.go

```diff
@@ -0,0 +1,200 @@
+package operations
+
+import (
+	"context"
+	"encoding/json"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"runtime"
+	"sort"
+	"strings"
+
+	"github.com/pkg/errors"
+	"github.com/rclone/rclone/fs"
+	"github.com/rclone/rclone/fs/config"
+)
+
+// Creates an OptionResume that will be passed to Put/Upload
+func createResumeOpt(ctx context.Context, f fs.Fs, remote string, src fs.Object) (resumeOpt *fs.OptionResume) {
+	ci := fs.GetConfig(ctx)
+	resumeOpt = &fs.OptionResume{ID: "", Pos: 0, SetID: createSetID(ctx, f, remote, src)}
+	if ci.ResumeLarger >= 0 {
+		root := f.Root()
+		if runtime.GOOS == "windows" {
+			if root[:4] == "//?/" {
+				root = root[4:]
+			}
+			if root[1] == ':' {
+				root = strings.Replace(root, ":", "", 1)
+			}
+		}
+		cacheName := filepath.Join(config.CacheDir, "resume", f.Name(), root, remote)
+		resumeID, hashName, hashState, attemptResume := readResumeCache(ctx, f, src, cacheName)
+		if attemptResume {
+			fs.Debugf(f, "Existing resume cache file found: %s. A resume will now be attmepted.", cacheName)
+			position, resumeErr := f.Features().Resume(ctx, remote, resumeID, hashName, hashState)
+			if resumeErr != nil {
+				fs.Errorf(src, "Resume canceled: %v", resumeErr)
+			} else if position > int64(ci.ResumeLarger) {
+				(*resumeOpt).Pos = position
+			}
+		}
+	}
+	return resumeOpt
+}
+
+// SetID will be called by backend's Put/Update function if the object's upload
+// could be resumed upon failure
+//
+// SetID takes the passed resume ID, hash state, hash name and Fingerprint of the object and stores it in
+// --cache-dir so that future Copy operations can resume the upload if it fails
+func createSetID(ctx context.Context, f fs.Fs, remote string, src fs.Object) func(ID, hashName, hashState string) (err error) {
+	ci := fs.GetConfig(ctx)
+	cacheCleaned := false
+	return func(ID, hashName, hashState string) (err error) {
+		rootCacheDir := filepath.Join(config.CacheDir, "resume")
```

Comment:
@mcalman 
Sorry. I thought I've finished reviewing but just now I noticed that this place violates a Go convention.
The context pointer "ctx" must always be passed only with call stack, but here it's passed from the resume option setup code in `createSetID` via closure to the returned anonymus function that will be called later by backend, probably in a separate call stack.
I suggest to clarify this place as follows: get rid of the anonymous function with its work data passed via closure. Instead, declare it as a public method of the resume option structure:
```
func (opt *fs.OptionResume) SetID(ctx context.Context, ID, hashName,  hashState string) error {
 ...
}
```
Then call it in backend as `err := resumeOpt.SetID(ctx, id, ...)`.
Pass its state, e.g `cacheCleaned` as a private member of the resume option structure.
Don't pass context or anything derived from it (`ci`) via structure, it's a violation altogether.
Derive it righr from new context in `SetID`.